src/librustc/infer/canonical.rs

```diff
@@ -0,0 +1,988 @@
+    /// canonical inputs to your query.
+    pub fn instantiate_canonical_with_fresh_inference_vars<T>(
+        &self,
+        span: Span,
+        canonical: &Canonical<'tcx, T>,
+    ) -> (T, CanonicalVarValues<'tcx>)
+    where
+        T: TypeFoldable<'tcx>,
+    {
+        let canonical_inference_vars =
+            self.fresh_inference_vars_for_canonical_vars(span, canonical.variables);
+        let result = canonical.substitute(self.tcx, &canonical_inference_vars);
+        (result, canonical_inference_vars)
+    }
+
+    /// Given the "infos" about the canonical variables from some
+    /// canonical, creates fresh inference variables with the same
+    /// characteristics. You can then use `substitute` to instantiate
+    /// the canonical variable with these inference variables.
+    pub fn fresh_inference_vars_for_canonical_vars(
+        &self,
+        span: Span,
+        variables: &Slice<CanonicalVarInfo>,
+    ) -> CanonicalVarValues<'tcx> {
+        let var_values: IndexVec<CanonicalVar, Kind<'tcx>> = variables
+            .iter()
+            .map(|info| self.fresh_inference_var_for_canonical_var(span, *info))
+            .collect();
+
+        CanonicalVarValues { var_values }
+    }
+
+    /// Given the "info" about a canonical variable, creates a fresh
+    /// inference variable with the same characteristics.
+    pub fn fresh_inference_var_for_canonical_var(
+        &self,
+        span: Span,
+        cv_info: CanonicalVarInfo,
+    ) -> Kind<'tcx> {
+        match cv_info.kind {
+            CanonicalVarKind::Ty(ty_kind) => {
+                let ty = match ty_kind {
+                    CanonicalTyVarKind::General => {
+                        self.next_ty_var(
+                            // FIXME(#48696) this handling of universes is not right.
+                            ty::UniverseIndex::ROOT,
+                            TypeVariableOrigin::MiscVariable(span),
+                        )
+                    }
+
+                    CanonicalTyVarKind::Int => self.tcx.mk_int_var(self.next_int_var_id()),
+
+                    CanonicalTyVarKind::Float => self.tcx.mk_float_var(self.next_float_var_id()),
+                };
+                Kind::from(ty)
+            }
+
+            CanonicalVarKind::Region => {
+                Kind::from(self.next_region_var(RegionVariableOrigin::MiscVariable(span)))
+            }
+        }
+    }
+
+    /// Given the (canonicalized) result to a canonical query,
+    /// instantiates the result so it can be used, plugging in the
+    /// values from the canonical query. (Note that the result may
+    /// have been ambiguous; you should check the certainty level of
+    /// the query before applying this function.)
+    ///
+    /// It's easiest to explain what is happening here by
+    /// example. Imagine we start out with the query `?A: Foo<'static,
+    /// ?B>`. We would canonicalize that by introducing two variables:
+    ///
+    ///     ?0: Foo<'?1, ?2>
```

Comment:
@nikomatsakis 
> But also, that string does not appear in syntax_ext

It appears here

Reply:
Yes, I now, thanks =) I just thought we didn't test examples in librustc (imo, we still shouldn't, but maybe we changed that?)

Reply:
Didn't notice your commit before I commented :)

Reply:
no worries =) in any case, I think maybe we still try to *parse* even if we don't run tests...

Reply:
(for some weird reason `./x.py test` is also giving me other strange, doc-test-related errors locally...)

Reply:
Couldn't you have added `ignore`? I think I've seen this in librustc recently.

Reply:
> other strange, doc-test-related errors

If it’s `E0003` in the error index, try `rm build/tmp/extended-error-metadata/x86_64-unknown-linux-gnu/librustc_const_eval.json`